### PR TITLE
[XrdHttp] Allow XrdHttp to provide the values of multiple headers

### PR DESCRIPTION
**Pull request by bbockelm**
In HTTP, headers can be provided multiple times.  However, in our current header-to-CGI translation, only one copy of the value makes it through (the query parameter string is successfully constructed with multiple parameters but this is later put into a hash inside XRootD).
    
This provides a new flag to the header2cgi command, `-multi`, that will translate multiple values to a comma-separated string.  So if we set:

```
http.header2cgi -multi Authorization authz
```

in the configuration then the following headers in a request:

```
Authorization: foo
Authorization: bar
```

would become the query string `?authz=foo,bar`.

Note the obfuscation routine is tweaked to allow `,` to be a character expected in the token.  This allows multiple authorization headers to be correctly obfuscated.

### CODE REVIEW

**Comment by ccaffy** [comment]
Thanks @amadio, very much appreciated!

File: src/XrdHttp/XrdHttpReq.hh [hunk 1]
```diff
@@ -164,18 +164,9 @@ private:
...
-    depth = 0;
-    opaque = 0;
-    writtenbytes = 0;
-    fopened = false;
-    headerok = false;
-    mScitag = -1;
-  };
+    prot(protinstance),
+    readRangeHandler(rcfg)
+  {}
```
Comment (amadio) [inline src/XrdHttp/XrdHttpReq.hh:170]: Not sure if others agree, but I find the old way easier to understand. I'd do
```suggestion
    prot(protinstance),
    readRangeHandler(rcfg),
    length(0),
    depth(0),
    opaque(nullptr),
    writtenbytes(0),
    fopened(false),
    headerok(false),
    mScitag(-1)
  {}
```
and drop the changes on the members directly.

File: src/XrdHttp/XrdHttpProtocol.hh [hunk 2]
```diff
@@ -436,6 +474,10 @@ protected:
   /// Rules that turn HTTP headers to cgi tokens in the URL, for internal comsumption
   static std::map< std::string, std::string > hdr2cgimap;
 
+  /// Rules that turn HTTP headers to cgi tokens in the URL; if the header is given multiple times, they
+  /// will be added to the CGI in a comma-separate manner.
+  static detail::HeaderMap hdr2cgimultimap;
```
Comment (amadio) [inline src/XrdHttp/XrdHttpProtocol.hh:479]: Are these `static` variables declared in a header correct? These look like they should have been declared as `static inline` to not have a copy in each translation unit.

File: tests/XrdOucTests/XrdOucUtilsTests.cc [hunk 3]
```diff
@@ -103,7 +103,7 @@ static const std::string authz_strings[] = {
   "/path/test.txt?scitag.flow=44&authz=REDACTED done close.",
   "/path/test.txt?authz=REDACTED&scitag.flow=44 done close.",
   "(message: kXR_stat (path: /tmp/xrootd/public/foo?authz=REDACTED&pelican.timeout=3s, flags: none) ).",
-  "(message: kXR_stat (path: /tmp/xrootd/public/foo?pelican.timeout=3s&authz=REDACTED, flags: none) ).",
+  "(message: kXR_stat (path: /tmp/xrootd/public/foo?pelican.timeout=3s&authz=REDACTED flags: none) ).",
```
Comment (amadio) [inline tests/XrdOucTests/XrdOucUtilsTests.cc:106]: This is a consequence of the unrelated change before. The test is correct and cannot be changed. Please remove this modification from the pull request.

**Review by amadio — CHANGES_REQUESTED** [review]

File: src/XrdOuc/XrdOucUtils.cc [hunk 4]
```diff
@@ -1427,7 +1443,7 @@ static bool is_token_character(int c)
   if (isalnum(c))
     return true;
 
-  static constexpr char token_chars[] = "-._~+/=:";
+  static constexpr char token_chars[] = "-._~+/=:,";
```
[the remark this answers is not in the document]
Comment (amadio) [reply src/XrdOuc/XrdOucUtils.cc:1446]: @bbockelm Sorry, I missed the explanation in the description. However, we cannot add comma here. Headers should be in their own lines, I don't see why adding comma is necessary.

File: src/XrdHttp/XrdHttpReq.hh (hunk 1 above)
Comment (ccaffy) [reply src/XrdHttp/XrdHttpReq.hh:170]: +1 yes, it's easier to see to what values the variables are assigned to by default in the constructor.

File: src/XrdOuc/XrdOucUtils.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (bbockelm) [reply src/XrdOuc/XrdOucUtils.cc:1446]: It's necessary because redaction isn't redacting only headers, it's also redacting CGI where tokens may show up.

So, if you have:

```
Authorization: foo
Authorization: bar
```

then the corresponding cgi is:

```
authz=foo,bar
```

Without this change, the logging is:

```
authz=REDACTED,bar
```

leaking the token `bar` into the log.  With the change, you have:

```
authz=REDACTED
```

[the remark this answers is not in the document]
Comment (ccaffy) [reply src/XrdOuc/XrdOucUtils.cc:1446]: Yes, correct. But the problem is for the redaction of log lines, sometimes we see message like the one you saw in the unit test: `"(message: kXR_stat (path: /tmp/xrootd/public/foo?pelican.timeout=3s&authz=REDACTED, flags: none) ).",`
 which means that if we obfuscate the full line considering `,` is a valid token characeter, then we lose some information on the logs (`flags: none`).

File: src/XrdHttp/XrdHttpProtocol.hh (hunk 2 above)
Comment (bbockelm) [reply src/XrdHttp/XrdHttpProtocol.hh:479]: Is that true?  From https://en.cppreference.com/w/cpp/language/static, it states that statics are external linkage (meaning you have to explicitly initialize them in one of the translation units, which is done in `XrdHttpProtocol.cc`).

I'm not an expert in this area of C++ but copying initializing a member into each translation unit that imports a header would be quite different from my understanding of statics.  If you have any links for me to read -- since I see `static inline` was introduced in C++17 -- I'd appreciate it!

**Comment by amadio** [comment]
These changes seem to go against what is in [RFC 7230](https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.2). Combined with [RFC 7235](https://datatracker.ietf.org/doc/html/rfc7235#section-4.2), which says `Authorization:` is not a comma separated list, it seems that what's introduced here is not going to work well.

I think it's better to use another header name which is specific to Pelican (like `X-Pelican-Auth...`) if you want to do things which don't conform with the RFCs.

File: src/XrdOuc/XrdOucUtils.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (amadio) [reply src/XrdOuc/XrdOucUtils.cc:1446]: I think the conclusion from looking at the RFCs is that `Authorization:` should not appear more than once. Adding comma as valid token character is not great. How to handle these non-standard `authz` cgi elements in any case?

File: src/XrdHttp/XrdHttpReq.hh (hunk 1 above)
Comment (amadio) [reply src/XrdHttp/XrdHttpReq.hh:170]: > Regardless, I think either my or Guilherme's approach is reasonable -- the original code version of having it as statements in the constructor body is not.

I agree that the statements inside the constructor are not nice. They make it [non-trivial](https://en.cppreference.com/w/cpp/language/default_constructor#Trivial_default_constructor), for instance, which we'd like to avoid.

File: src/XrdOuc/XrdOucUtils.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (bbockelm) [reply src/XrdOuc/XrdOucUtils.cc:1446]: > How to handle these non-standard authz cgi elements in any case?

Indeed, that's part of the next step and why I was working on #2381 in the first place.  For Pelican, we need to authenticate the cache and the client separately -- ensuring that an origin can ensure that (a) this is a valid request on behalf of a client and (b) the data is going to a known cache.  This prevents unregistered caches (or rogue clients) from accessing the origin directly.

That work is still on a branch (https://github.com/bbockelm/xrootd/tree/multi_token_authz) that isn't quite functional enough for even a draft -- however, getting both tokens to the plugin is a prerequisite.

[the remark this answers is not in the document]
Comment (amadio) [reply src/XrdOuc/XrdOucUtils.cc:1446]: For caches, I think it's better to treat them like proxies, and use a [`Proxy-Authorization`](https://datatracker.ietf.org/doc/html/rfc7235#section-4.4) header rather than repeat the `Authorization:` header, which is not conforming to the standards. Or you can create `X-Pelican*` headers which can handle this outside of XRootD. I'd like to hear @abh3's and @osschar's opinions on this, but I hope they will agree that we want to avoid having non-standard behavior in XRootD, so that it can interoperate well with other software.

[the remark this answers is not in the document]
Comment (abh3) [reply src/XrdOuc/XrdOucUtils.cc:1446]: I agree with @amadio that we really should avoid non-standard use of headers , especially when we are co-opting a standard header to do our bidding. In this particular case, I recall @bbockelm and I had a discussion on how to solve this particular problem by using RFC 7800 to generate special tokens that Xcache could use to access origins on behalf of clients (assuming you trust Xcache to do a proper client credential check - which we already do for x509). It appears that seems to have veered off course likely because I didn't produce the proposal to go that way. My other problem with multi-token transactions is that the framing overhead is moving in the wrong direction. We've already had problems with excessively large tokens and needed to make some code changes to accommodate them, This just makes tokens larger by compounding them. That said, if we need this kind of flexibility and if it is specific to Pelican, then either a special X-Pelican header or more generic special header, if this is useful beyond Pelican, would be warranted here. That doesn't address the obfuscation problem but I've already had a discussion with @amadio on a different mechanism of obfuscation that is insensitive to the token character set, which should drive the final nail into that coffin.

**Comment by bbockelm** [comment]
> I think it's better to use another header name which is specific to Pelican (like X-Pelican-Auth...) if you want to do things which don't conform with the RFCs.

Since the RFC's you cite are marked as obsolete, I went digging through the RFC that they were obsoleted by, RFC9110.

Here's the grammar for the Authorization header:
```
  token          = 1*tchar

  tchar          = "!" / "#" / "$" / "%" / "&" / "'" / "*"
                 / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~"
                 / DIGIT / ALPHA
                 ; any VCHAR, except delimiters
  OWS            = *( SP / HTAB )
                 ; optional whitespace
  obs-text       = %x80-FF
  quoted-pair    = "\" ( HTAB / SP / VCHAR / obs-text )
  qdtext         = HTAB / SP / %x21 / %x23-5B / %x5D-7E / obs-text
  quoted-string  = DQUOTE *( qdtext / quoted-pair ) DQUOTE
  BWS            = OWS
                 ; "bad" whitespace
auth-scheme = token
auth-param = token BWS "=" BWS ( token / quoted-string )
credentials = auth-scheme [ 1*SP ( token68 / [ auth-param *( OWS ","
 OWS auth-param ) ] ) ]
Authorization = credentials
```

So, while the `Authorization` header itself is not comma-separated, a comma is a valid character in the header.

(Another interesting find while going through the OAuth 2.0 RFC: the RFC-compliant way to include the token in the URL query parameters is to use the `access_token` key, not `authz`)

For Pelican, I don't think it'd be good to use a different header as there's special security handling for `Authorization` in the browser.  I'd be willing to use a custom auth-scheme though:

```
Authorization: Pelican user_token=abcd,cache_token=efgh
```

However, I need the tokens to be parsed by the `XrdSciTokens` module: I really don't want to fork the scitokens module and have all my improvements stay bottled up in Pelican-land.

**Comment by bbockelm** [comment]
After a night's reflection, my last comment provides a way forward:
1. For the `root://` protocol, ZTN + URL-based token already provides two different tokens.
2. For `https://`, if we add `access_token` for RFC correctness then that plus the `authz` header suffices.  Like @amadio suggests, I can map a custom header (e.g., `X-Pelican-Authorization`) to `access_token`.  Because Pelican scrubs all log lines, I don't need to worry about the fact that XRootD doesn't scrub the `X-Pelican-Authorization` line.

I posted the resulting work as #2397 which achieves the desired functionality (including unit tests!) without this particular PR.

While this may still be interesting to someone out there, I don't want to merge things without a clear use case.  Hence, I'm going to close the PR but keep the branch around in case someone ever wants to revive it.
